src-java/kilda-model/src/main/java/org/openkilda/model/FlowEndpoint.java

```diff
@@ -16,48 +16,109 @@
     @JsonProperty("outer_vlan_id")
-    private final int vlanId;
+    private final int outerVlanId;
+
+    @JsonProperty("inner_vlan_id")
+    private final int innerVlanId;
 
     @JsonProperty("track_lldp_connected_devices")
     private final boolean trackLldpConnectedDevices;
 
     @JsonProperty("track_arp_connected_devices")
     private final boolean trackArpConnectedDevices;
 
-    public FlowEndpoint(SwitchId switchId, Integer portNumber) {
+    public FlowEndpoint(SwitchId switchId, int portNumber) {
         this(switchId, portNumber, 0);
     }
 
-    public FlowEndpoint(SwitchId switchId, Integer portNumber, int vlanId) {
-        this(switchId, portNumber, vlanId, false, false);
+    public FlowEndpoint(SwitchId switchId, int portNumber, int outerVlanId) {
+        this(switchId, portNumber, outerVlanId, 0, false, false);
+    }
+
+    public FlowEndpoint(SwitchId switchId, int portNumber, int outerVlanId, int innerVlanId) {
+        this(switchId, portNumber, outerVlanId, innerVlanId, false, false);
     }
 
     @JsonCreator
     @Builder(toBuilder = true)
     public FlowEndpoint(
             @JsonProperty("switch_id") SwitchId switchId,
             @JsonProperty("port_number") Integer portNumber,
-            @JsonProperty("outer_vlan_id") int vlanId,
+            @JsonProperty("outer_vlan_id") int outerVlanId,
+            @JsonProperty("inner_vlan_id") int innerVlanId,
             @JsonProperty("track_lldp_connected_devices") boolean trackLldpConnectedDevices,
             @JsonProperty("track_arp_connected_devices") boolean trackArpConnectedDevices) {
         super(switchId, portNumber);
-        this.vlanId = vlanId;
+
         this.trackLldpConnectedDevices = trackLldpConnectedDevices;
         this.trackArpConnectedDevices = trackArpConnectedDevices;
+
+        // normalize VLANs representation
+        List<Integer> vlanStack = makeVlanStack(innerVlanId, outerVlanId);
+        if (1 < vlanStack.size()) {
+            this.outerVlanId = vlanStack.get(1);
+            this.innerVlanId = vlanStack.get(0);
+        } else if (!vlanStack.isEmpty()) {
+            this.outerVlanId = vlanStack.get(0);
+            this.innerVlanId = 0;
+        } else {
+            this.outerVlanId = 0;
+            this.innerVlanId = 0;
+        }
+    }
+
+    @JsonIgnore
+    public List<Integer> getVlanStack() {
+        return makeVlanStack(innerVlanId, outerVlanId);
+    }
+
+    /**
+     * Scan provided sequence for valid VLAN IDs and return them as a list.
+     */
+    public static List<Integer> makeVlanStack(Integer... sequence) {
+        return Stream.of(sequence)
+                .filter(FlowEndpoint::isVlanIdSet)
+                .collect(Collectors.toList());
     }
 
     public static boolean isVlanIdSet(Integer vlanId) {
         return vlanId != null && 0 < vlanId;
     }
+
+    @Override
+    public String toString() {
+        String view = String.format("switchId=\"%s\" port=%d", switchId, portNumber);
```

Comment:
You mix 2 concat practice here, implicit use of StringBuilder and string.format. Chose one.

Reply:
Done.